File: streamlit-chatbot-app/dev-proxy.py
#!/usr/bin/env python3
"""
Development proxy server for testing chatbot with live AgentCore runtime
Uses boto3 to call bedrock-agentcore API directly
"""

import json
import os
import boto3
from flask import Flask, request, jsonify
from flask_cors import CORS

app = Flask(__name__)
CORS(app)  # Enable CORS for local development

# Your AgentCore runtime configuration
AGENT_RUNTIME_ARN = os.getenv(
    'AGENT_RUNTIME_ARN',
    'arn:aws:bedrock-agentcore:us-east-1:006262944085:runtime/strands_agent_agentcore_runtime-H37svwDSIJ'
)
AWS_REGION = os.getenv('AWS_REGION', 'us-east-1')

# Initialize Bedrock AgentCore client
bedrock_agentcore = boto3.client('bedrock-agentcore', region_name=AWS_REGION)

@app.route('/health', methods=['GET'])
def health():
    """Health check endpoint"""
    return jsonify({
        'status': 'ok',
        'agentArn': AGENT_RUNTIME_ARN,
        'region': AWS_REGION
    })

@app.route('/chat', methods=['POST'])
def chat():
    """Chat endpoint that invokes AgentCore runtime"""
    try:
        data = request.get_json()
        message = data.get('message', '')
        session_id = data.get('sessionId', f'dev-session-{int(__import__("time").time())}')

        if not message:
            return jsonify({'error': 'Message is required'}), 400

        print(f"\n📩 Received message: \"{message}\"")
        print(f"🔑 Session ID: {session_id}")
        print(f"🤖 Agent Runtime ARN: {AGENT_RUNTIME_ARN}")
        print(f"⏳ Invoking agent...")

        # Generate a proper session ID (min 33 chars)
        if not session_id or len(session_id) < 33:
            import uuid
            session_id = f"session-{uuid.uuid4()}"

        # Invoke AgentCore runtime with correct parameters
        # Payload must be JSON bytes with 'prompt' key
        payload_dict = {'prompt': message}
        payload_bytes = json.dumps(payload_dict).encode('utf-8')

        response = bedrock_agentcore.invoke_agent_runtime(
            agentRuntimeArn=AGENT_RUNTIME_ARN,
            runtimeSessionId=session_id,
            payload=payload_bytes,
            contentType='application/json',
            accept='application/json'
        )

        print(f"✅ Response received")
        print(f"Response keys: {response.keys()}")

        # Extract response from the streaming response body
        agent_response = ""

        if 'response' in response:
            # Read the streaming body
            streaming_body = response['response']
            agent_response = streaming_body.read().decode('utf-8')
            print(f"📄 Agent response: {agent_response}")
        elif 'completion' in response:
            # Handle streaming response (alternative format)
            for event in response['completion']:
                print(f"Event: {event}")
                if 'chunk' in event:
                    chunk = event['chunk']
                    if 'bytes' in chunk:
                        chunk_text = chunk['bytes'].decode('utf-8')
                        agent_response += chunk_text
                        print(chunk_text, end='', flush=True)

        print(f"\n✅ Agent response complete ({len(agent_response)} chars)\n")

        return jsonify({
            'response': agent_response or 'No response from agent',
            'sessionId': session_id
        })

    except Exception as e:
        import traceback
        print(f"❌ Error: {e}", flush=True)
        print(f"Error type: {type(e).__name__}", flush=True)
        print(f"Traceback:", flush=True)
        traceback.print_exc()

        return jsonify({
            'error': 'Internal server error',
            'details': str(e),
            'type': type(e).__name__
        }), 500

if __name__ == '__main__':
    print("""
╔════════════════════════════════════════════════════════════╗
║  🚀 Development Proxy Server Running (Python)              ║
╟────────────────────────────────────────────────────────────╢
║  URL:        http://localhost:3001                         ║
║  Endpoint:   http://localhost:3001/chat                    ║
║  Health:     http://localhost:3001/health                  ║
║                                                            ║
║  Agent ARN:  {}... ║
║  Region:     {}                                      ║
╟────────────────────────────────────────────────────────────╢
║  Using AWS credentials from your local profile             ║
╚════════════════════════════════════════════════════════════╝
    """.format(AGENT_RUNTIME_ARN[:60], AWS_REGION))

    app.run(host='localhost', port=3001, debug=True)

File: streamlit-chatbot-app/dev-proxy.js
/**
 * Development proxy server for testing chatbot with live AgentCore runtime
 * Uses AWS SDK to call bedrock-agentcore-runtime API
 */

import express from 'express';
import cors from 'cors';
import { fromNodeProviderChain } from '@aws-sdk/credential-providers';
import { SignatureV4 } from '@smithy/signature-v4';
import { HttpRequest } from '@smithy/protocol-http';
import { Sha256 } from '@aws-crypto/sha256-js';

const app = express();
const PORT = 3001;

// Your AgentCore runtime configuration
const AGENT_RUNTIME_ARN = process.env.AGENT_RUNTIME_ARN || 'arn:aws:bedrock-agentcore:us-east-1:006262944085:runtime/strands_agent_agentcore_runtime-H37svwDSIJ';
const AWS_REGION = process.env.AWS_REGION || 'us-east-1';

// Enable CORS for local development
app.use(cors());
app.use(express.json());

// Health check endpoint
app.get('/health', (req, res) => {
  res.json({ status: 'ok', agentArn: AGENT_RUNTIME_ARN, region: AWS_REGION });
});

// Chat endpoint
app.post('/chat', async (req, res) => {
  try {
    const { message, sessionId } = req.body;

    if (!message) {
      return res.status(400).json({ error: 'Message is required' });
    }

    console.log(`\n📩 Received message: "${message}"`);
    console.log(`🔑 Session ID: ${sessionId}`);
    console.log(`🤖 Agent Runtime ARN: ${AGENT_RUNTIME_ARN}`);

    // Get AWS credentials
    const credentialsProvider = fromNodeProviderChain();
    const credentials = await credentialsProvider();

    // Prepare request - using bedrock-agentcore service
    const endpoint = `https://bedrock-agentcore.${AWS_REGION}.amazonaws.com`;
    const path = `/invoke-agent-runtime`;

    const requestBody = JSON.stringify({
      runtimeArn: AGENT_RUNTIME_ARN,
      runtimeSessionId: sessionId || `dev-session-${Date.now()}`,
      inputText: message,
    });

    // Create HTTP request
    const request = new HttpRequest({
      method: 'POST',
      protocol: 'https:',
      hostname: `bedrock-agentcore.${AWS_REGION}.amazonaws.com`,
      path: path,
      headers: {
        'Content-Type': 'application/json',
        'Host': `bedrock-agentcore.${AWS_REGION}.amazonaws.com`,
      },
      body: requestBody,
    });

    // Sign request with SigV4
    const signer = new SignatureV4({
      credentials: credentials,
      region: AWS_REGION,
      service: 'bedrock-agentcore',
      sha256: Sha256,
    });

    const signedRequest = await signer.sign(request);

    console.log(`⏳ Invoking agent at ${endpoint}${path}...`);

    // Make the request
    const response = await fetch(`${endpoint}${path}`, {
      method: signedRequest.method,
      headers: signedRequest.headers,
      body: signedRequest.body,
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error(`❌ API Error (${response.status}):`, errorText);
      throw new Error(`AgentCore API error: ${response.status} - ${errorText}`);
    }

    // Read response body
    const responseData = await response.json();
    console.log(`✅ Response:`, responseData);

    // Extract agent response
    const agentResponse = responseData.output || responseData.completion || 'No response from agent';

    res.json({
      response: agentResponse,
      sessionId: sessionId,
    });

  } catch (error) {
    console.error('❌ Error:', error);

    res.status(500).json({
      error: 'Internal server error',
      details: error.message,
      name: error.name,
    });
  }
});

app.listen(PORT, () => {
  console.log(`
╔════════════════════════════════════════════════════════════╗
║  🚀 Development Proxy Server Running                       ║
╟────────────────────────────────────────────────────────────╢
║  URL:        http://localhost:${PORT}                         ║
║  Endpoint:   http://localhost:${PORT}/chat                    ║
║  Health:     http://localhost:${PORT}/health                  ║
║                                                            ║
║  Agent ARN:  ${AGENT_RUNTIME_ARN.substring(0, 60)}... ║
║  Region:     ${AWS_REGION}                                      ║
╟────────────────────────────────────────────────────────────╢
║  Using AWS credentials from your local profile             ║
╚════════════════════════════════════════════════════════════╝
  `);
});
